naicreport/glance/glance.go

```diff
@@ -0,0 +1,571 @@
+package glance
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+	"time"
+
+	"naicreport/jobstate"
+	"naicreport/sonalyze"
+	"naicreport/util"
+)
+
+// Parameters for what are "recent" events/averages and what are events "longer ago" and even "long
+// ago", these could be defaults that are overrideable but are currently hardcoded.
+//
+// Note that for the load report, where we have special switches for the accumulation interval, only
+// some values are acceptable: 30 min => --half-hourly, 60 min => --hourly, 12 hrs => --half-daily,
+// 24 hrs => --daily (and there's --none but let's not go there).
+
+const (
+	RECENT_MINS = 30
+	LONGER_MINS = 12 * 60
+	LONG_MINS   = 24 * 60
+)
+
+const nanosPerSec time.Duration = 1000000000
+
+func Report(progname string, args []string) error {
+	progOpts := util.NewStandardOptions(progname + " at-a-glance")
+	sonalyzePathPtr := progOpts.Container.String("sonalyze", "", "Path to sonalyze executable (required)")
+	configFilenamePtr := progOpts.Container.String("config-file", "", "Path to system config file (required)")
+	statePathPtr := progOpts.Container.String("state-path", "", "Path to directory holding database state (required)")
+	err := progOpts.Parse(args)
+	if err != nil {
+		return err
+	}
+
+	if len(progOpts.DataFiles) > 0 {
+		return errors.New("The -- filename ... option does not make sense to this module (yet)")
+	}
+
+	sonalyzePath, err := util.CleanPath(*sonalyzePathPtr, "-sonalyze")
+	if err != nil {
+		return err
+	}
+	configFilename, err := util.CleanPath(*configFilenamePtr, "-config-file")
+	if err != nil {
+		return err
+	}
+	statePath, err := util.CleanPath(*statePathPtr, "-state-path")
+	if err != nil {
+		return err
+	}
+
+	// It's possible that the code in this module would be a little less verbose if there were a
+	// centralized database mapping host -> glanceRecord that is just passed around for everyone to
+	// use.  But it's not obvious the code would be as easy to understand.
+
+	ujbh, err := collectUsersJobsAndHosts(sonalyzePath, RECENT_MINS, LONGER_MINS, progOpts)
+	if err != nil {
+		return err
+	}
+	sdbh, err := collectStatusData(sonalyzePath, progOpts)
+	if err != nil {
+		return err
+	}
+	labh, err := collectLoadAverages(sonalyzePath, configFilename, RECENT_MINS, LONGER_MINS, progOpts)
+	if err != nil {
+		return err
+	}
+	cpuhogFilename := path.Join(statePath, "cpuhog-state.csv")
+	cpuhogCutoff := time.Now().UTC().Add(time.Duration(-LONG_MINS) * 60 * nanosPerSec)
+	hogsbh, err := collectCpuhogs(cpuhogFilename, cpuhogCutoff)
+	if err != nil {
+		return err
+	}
+	deadweightFilename := path.Join(statePath, "deadweight-state.csv")
+	deadweightCutoff := time.Now().UTC().Add(time.Duration(-LONG_MINS) * 60 * nanosPerSec)
```

Comment:
Use the same timestamp as above!  First, get the current time only once.  Then, the two cutoffs are now the same, so we just compute that quantity once.